src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.Analyzer.cs

```diff
@@ -0,0 +1,351 @@
+                    {
+                        case SyntaxKind.ParenthesizedExpression:
+                        case SyntaxKind.CastExpression:
+                            // Parentheses and cast expressions do not contribute to the flow analysis.
+                            continue;
+
+                        case SyntaxKind.LogicalNotExpression:
+                            // The !-operator negates the definitive assignment state.
+                            defAssignedWhenTrue = !defAssignedWhenTrue;
+                            continue;
+
+                        case SyntaxKind.LogicalAndExpression:
+                            if (!defAssignedWhenTrue)
+                            {
+                                // Since the pattern variable is only definitely assigned if the pattern
+                                // succeeded, in the following cases it would not be safe to use pattern-matching.
+                                // For example:
+                                //
+                                //      if ((x = o as string) == null && SomeExpression)
+                                //      if ((x = o as string) != null || SomeExpression)
+                                //
+                                // Here, x would never be definitely assigned if pattern-matching were used.
+                                return false;
+                            }
+
+                            continue;
+
+                        case SyntaxKind.LogicalOrExpression:
+                            if (defAssignedWhenTrue)
+                            {
+                                // See the comment above.
+                                return false;
+                            }
+
+                            continue;
+
+                        case SyntaxKind.ConditionalExpression:
+                            var conditionalExpression = (ConditionalExpressionSyntax)current;
+                            if (LocalFlowsIn(defAssignedWhenTrue
+                                    ? conditionalExpression.WhenFalse
+                                    : conditionalExpression.WhenTrue))
+                            {
+                                // In a conditional expression, the pattern variable
+                                // would not be definitely assigned in the opposite branch.
+                                return false;
+                            }
+
+                            return CheckExpression(conditionalExpression);
+
+                        case SyntaxKind.ForStatement:
+                            var forStatement = (ForStatementSyntax)current;
+                            if (!forStatement.Condition.Span.Contains(_comparison.Span))
+                            {
+                                // In a for-statement, only the condition expression
+                                // can make this definitely assigned in the loop body.
+                                return false;
```

Comment:
what about if it was in the for-loop initializer?

Reply:
Added a test to very it's missing. how it can be useful anyway?

Note: we still support a ternary in for-loop initializer and the allowed scope would be the ternary itself.